In [1]:
!pip install pysam
import pysam

     |████████████████████████████████| 15.0 MB 6.5 MB/s 


In [11]:
samfile = pysam.AlignmentFile("merged-tumor.bam", "rb")

i=0
for read in samfile:
    print(read)
    print ('Read '+str(i+1)+' flag: '+str(read.flag))
    i=i+1
    if i>10:
      break


samfile.close()

C0HVYACXX120402:7:1207:5722:57044	1187	#20	9483249	27	76M	#20	9483382	209	TTTTCAAACAGTATCTATGCCTGCCAAATGTGAACATATAAAAAAAAACCAGAATGTGCCATTCTGATTTAAACTG	array('B', [28, 28, 27, 29, 31, 30, 31, 31, 29, 31, 35, 30, 29, 31, 34, 30, 29, 23, 41, 32, 20, 30, 29, 34, 34, 29, 30, 31, 30, 30, 30, 33, 33, 26, 39, 12, 25, 19, 32, 30, 35, 28, 35, 33, 23, 33, 35, 36, 30, 38, 33, 41, 34, 35, 31, 33, 23, 30, 30, 36, 27, 32, 29, 34, 35, 41, 33, 31, 33, 29, 32, 32, 31, 31, 31, 34])	[('XA', 'GL000217.1,-110754,76M,1;'), ('BD', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('MD', '76'), ('RG', '1'), ('BI', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('NM', 0), ('MQ', 27), ('AS', 76), ('XS', 71)]
Read 1 flag: 1187
D0RE2ACXX120401:1:2105:2631:113383	163	#20	9483249	27	76M	#20	9483382	209	TTTTCAAACAGTATCTATGCCTGCCAAATGTGAACATATAAAAAAAAACCAGAATGTGCCATTCTGATTTAAACTG	array('B', [28, 29, 27, 30, 31, 30, 31, 31, 34, 31, 40, 30, 29, 34, 41, 34,


##The flag for the first read is 1187, which means that these bits are set:



*   read paired (0x1)
*   read mapped in proper pair (0x2)
*   mate reverse strand (0x20)
*   second in pair (0x80)
*   read is PCR or optical duplicate (0x400)



##The flag for the second read is 163, which means that these bits are set:

*  read paired (0x1)
*  read mapped in proper pair (0x2)
*  mate reverse strand (0x20)
*  second in pair (0x80)

##The flag for the third read is 99, which means that these bits are set:
*  read paired (0x1)
*  read mapped in proper pair (0x2)
*  mate reverse strand (0x20)
*  first in pair (0x40)

---



In [13]:
samfile = pysam.AlignmentFile("merged-tumor.bam", "rb")

cnt_unmapped=0
cnt=0
cnt_mapping_quality_0=0
quality_sum=0
for read in samfile:
    if read.is_unmapped:
         cnt_unmapped += 1
    if read.mapping_quality==0:
         cnt_mapping_quality_0 += 1
    quality_sum=quality_sum+read.mapping_quality
    cnt += 1

print('Number of unmapped reads:')
print(cnt_unmapped)
print('Total number of reads:')
print(cnt)
print('Number of reads with mapping quality 0:')
print(cnt_mapping_quality_0)
print('Average mapping quality for all the reads:')
if cnt>0:
  print(quality_sum/cnt)
else:
  print(0)
print('Average mapping quality if reads with 0 mapp quality are filtered out:')
if cnt!=cnt_mapping_quality_0:
  print(quality_sum/(cnt-cnt_mapping_quality_0))
else:
  print(0)

samfile.close()

Number of unmapped reads:
17765
Total number of reads:
2921629
Number of reads with mapping quality 0:
126628
Average mapping quality for all the reads:
55.91379158681681
Average mapping quality if reads with 0 mapp quality are filtered out:
58.446975510921106
